worker/sshserver/server_test.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshserver_test
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"net"
+	"time"
+
+	"github.com/juju/testing"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc/test/bufconn"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/worker/sshserver"
+)
+
+type sshServerSuite struct {
+	testing.IsolationSuite
+
+	userSigner  ssh.Signer
+	jumpHostKey string
+}
+
+var _ = gc.Suite(&sshServerSuite{})
+
+func (s *sshServerSuite) SetUpSuite(c *gc.C) {
+	// Setup user signer
+	userKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	c.Assert(err, gc.IsNil)
+
+	userSigner, err := ssh.NewSignerFromKey(userKey)
+	c.Assert(err, gc.IsNil)
+
+	s.userSigner = userSigner
+
+	// Setup jump host private key
+	jumpHostKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	c.Assert(err, gc.IsNil)
+
+	jumpHostKeyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(jumpHostKey),
+	})
+
+	s.jumpHostKey = string(jumpHostKeyPEM)
+
+}
+
+func (s *sshServerSuite) TestSSHServer(c *gc.C) {
+	// Firstly, start the server on an in-memory listener
+	listener := bufconn.Listen(8 * 1024)
+	server, err := sshserver.NewSSHServer(
+		nil,
+		s.jumpHostKey,
+	)
+	c.Assert(err, gc.IsNil)
+
+	go func() {
+		err := server.Serve(listener)
+		c.Assert(err, gc.IsNil)
+	}()
+
+	// Dial the in-memory listener
+	conn, err := listener.Dial()
+	c.Assert(err, gc.IsNil)
+
+	// Open a client connection
+	jumpConn, _, terminatingReqs, err := ssh.NewClientConn(
+		conn,
+		"",
+		&ssh.ClientConfig{
+			HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+			Auth: []ssh.AuthMethod{
+				ssh.Password(""), // No password needed
+			},
+		},
+	)
+	c.Assert(err, gc.IsNil)
+	go ssh.DiscardRequests(terminatingReqs)
+
+	// Open direct-tcpip channel (the -J part of SSH)
+	d := struct {
+		DestAddr string
+		DestPort uint32
+		SrcAddr  string
+		SrcPort  uint32
+	}{
+		DestAddr: "1.postgresql.8419cd78-4993-4c3a-928e-c646226beeee.juju.local",
+		DestPort: 20,
+		SrcAddr:  "localhost",
+		SrcPort:  0,
+	}
+
+	jumpChannel, reqs, err := jumpConn.OpenChannel("direct-tcpip", ssh.Marshal(&d))
+	c.Assert(err, gc.IsNil)
+	go ssh.DiscardRequests(reqs)
+
+	// Now with this opened direct-tcpip channel, open a session connection
+	terminatingClientConn, terminatingClientChan, terminatingReqs, err := ssh.NewClientConn(
+		&channelConn{jumpChannel},
```

Comment:
It's unfortunate we need to wrap this jumpChannel in order to satisfy net.Conn. Because we'll (potentially?) be doing this same thing from the `juju ssh` command if we don't shell out to OpenSSH.

Reply:
Exactly, and kind of frustrating, I don't see another method to accept just the channel either. I did actually make a func which takes the channel but it felt a bit OTT and reverted to wrapping the channel again in the end...

Reply:
I figured out how to simplify this, use the initial `conn` with `ssh.NewClientConn` and then pass the 3 resulting values into ssh.NewClient() as below then use `ssh.Dial()` which does a jump.
```
	client := ssh.NewClient(jumpConn, chans, terminatingReqs)
	tunnel, err := client.Dial("tcp", "1.postgresql.8419cd78-4993-4c3a-928e-c646226beeee.juju.local:20")
```